File: ametras_fast_import_addon/vue-app/src/importer/batchSizeAdapter.ts
/**
 * Adaptive batch size controller.
 * Shared between addon and standalone modes.
 *
 * Starts at the middle level (typically 10 rows) as a tradeoff:
 * - Not too small: only ~3 batches to ramp up to max on good data
 * - Not too large: a failed batch of 10 triggers fewer retry requests
 *   than a failed batch of 100
 *
 * Levels: [1, min(10, max), max] — starts at level 1 (middle).
 * Increase: after 30 consecutive successful rows, step up one level.
 * Decrease: after 3 consecutive concurrency failures, step down one level.
 * Timeout: immediately step down one level.
 */
export class BatchSizeAdapter {
  currentSize: number
  private levels: number[]
  private levelIndex: number
  private successfulRows: number = 0
  private consecutiveFailures: number = 0

  static readonly SUCCESS_THRESHOLD = 30
  static readonly FAILURE_THRESHOLD = 3

  constructor(maxSize: number = 100) {
    // Build levels: [1, min(10, max), max] deduplicated
    const raw = [1, Math.min(10, maxSize), maxSize]
    this.levels = [...new Set(raw)].sort((a, b) => a - b)
    // Start at middle level — balances warmup cost vs failed-batch cost
    this.levelIndex = Math.min(1, this.levels.length - 1)
    this.currentSize = this.levels[this.levelIndex]
  }

  /** Record successful rows and step back up after recovery from a step-down. */
  recordSuccess(rowCount: number): void {
    this.consecutiveFailures = 0
    if (this.levelIndex >= this.levels.length - 1) return // already at max
    this.successfulRows += rowCount
    if (this.successfulRows >= BatchSizeAdapter.SUCCESS_THRESHOLD) {
      this.levelIndex++
      this.currentSize = this.levels[this.levelIndex]
      this.successfulRows = 0
    }
  }

  /** Record a batch failure and step down after enough consecutive failures. */
  recordFailure(): void {
    this.successfulRows = 0
    this.consecutiveFailures++
    if (this.consecutiveFailures >= BatchSizeAdapter.FAILURE_THRESHOLD && this.levelIndex > 0) {
      this.levelIndex--
      this.currentSize = this.levels[this.levelIndex]
      this.consecutiveFailures = 0
    }
  }

  /**
   * Immediately step down one level — used for timeouts where the batch is clearly too large.
   * Returns true if the size actually changed (stepped down), false if already at minimum.
   */
  recordTimeout(): boolean {
    this.successfulRows = 0
    this.consecutiveFailures = 0
    if (this.levelIndex > 0) {
      this.levelIndex--
      this.currentSize = this.levels[this.levelIndex]
      return true
    }
    return false
  }

  /** True when the adapter is at the smallest possible batch size. */
  get isAtMinimum(): boolean {
    return this.levelIndex === 0
  }
}

File: ametras_fast_import_addon/models/import_engine/batch_size_adapter.py
"""
Adaptive batch-size controller for the import engine.

Port of 16.0's BatchSizeAdapter (TypeScript) to Python.

The adapter tracks import outcomes and adjusts batch size up or down across
three levels: [1, min(10, max_size), max_size].  Thresholds count *rows*,
not batches, so behaviour scales naturally with the configured batch size.
"""
from __future__ import annotations

from .constants import (
    BATCH_SIZE_SUCCESS_THRESHOLD,
    BATCH_SIZE_FAILURE_THRESHOLD,
)


class BatchSizeAdapter:
    """
    Three-level adaptive batch-size controller.

    Levels are [1, min(10, max_size), max_size] deduplicated and sorted.
    The adapter starts at the middle level (index 1, or 0 if max_size <= 1).

    Threshold semantics (row-based, not batch-based):
    - Step UP  when accumulated successful rows >= success_threshold.
    - Step DOWN when consecutive batch failures >= FAILURE_THRESHOLD.
    - Step DOWN immediately on timeout; optionally raise the success threshold
      for the next ramp-up to prevent too-quick re-escalation.
    """

    def __init__(self, max_size: int):
        levels = sorted(set([1, min(10, max_size), max_size]))
        self._levels = levels
        self._level = min(1, len(levels) - 1)   # middle rung (or 0 if only 1 level)
        self._success_rows = 0
        self._consecutive_failures = 0
        self._success_threshold = BATCH_SIZE_SUCCESS_THRESHOLD

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def current_size(self) -> int:
        return self._levels[self._level]

    @property
    def is_at_minimum(self) -> bool:
        return self._level == 0

    # ------------------------------------------------------------------
    # Outcome recording
    # ------------------------------------------------------------------

    def record_success(self, row_count: int) -> None:
        """
        Record successfully processed rows.

        Resets the consecutive-failure counter and accumulates rows toward
        the step-up threshold.  Steps up at most one level per call.
        """
        self._consecutive_failures = 0
        self._success_rows += row_count
        if (self._success_rows >= self._success_threshold
                and self._level < len(self._levels) - 1):
            self._level += 1
            self._success_rows = 0
            self._success_threshold = BATCH_SIZE_SUCCESS_THRESHOLD

    def record_failure(self) -> None:
        """
        Record a failed batch (not a timeout — for those use record_timeout).

        Steps down one level after BATCH_SIZE_FAILURE_THRESHOLD consecutive
        failures; resets the failure counter on step-down.
        """
        self._success_rows = 0
        self._consecutive_failures += 1
        if self._consecutive_failures >= BATCH_SIZE_FAILURE_THRESHOLD:
            self._consecutive_failures = 0
            if self._level > 0:
                self._level -= 1

    def record_timeout(
        self,
        success_threshold_after_timeout: int | None = None,
    ) -> bool:
        """
        Record a timeout and immediately step down one level.

        Args:
            success_threshold_after_timeout: If provided, overrides the
                success threshold for the next ramp-up (use
                STANDALONE_POST_TIMEOUT_SUCCESS_THRESHOLD = 300 to slow
                re-escalation after a timeout).

        Returns:
            True if the adapter stepped down, False if already at minimum.
        """
        if self._level == 0:
            return False
        self._level -= 1
        self._consecutive_failures = 0
        self._success_rows = 0
        if success_threshold_after_timeout is not None:
            self._success_threshold = success_threshold_after_timeout
        return True

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def reset(self) -> None:
        """Reset to the initial middle level.  Call between import runs."""
        self._level = min(1, len(self._levels) - 1)
        self._success_rows = 0
        self._consecutive_failures = 0
        self._success_threshold = BATCH_SIZE_SUCCESS_THRESHOLD
